### PR TITLE
Implementing > 64K cigar operators for BAM

### PR DESCRIPTION
**Pull request by yfarjoun**
### Description
This is based on @lh3 's PR #1003  but implements it using htsjdk constructs rather than the low-level implementations that were made there.

This PR implements the long-cigar solution that was agreed to in hts-spec. In the presence of CIGAR with >65535 operators, we move the real CIGAR to the CG tag and put a fake full clipping CIGAR (`<readLength>S<referenceLength>N`) in its place.

Also, many tests and checks were implemented.
### Checklist

- [ ] Code compiles correctly
- [ ] New tests covering changes and new functionality
- [ ] All tests passing
- [ ] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

### CODE REVIEW

**Comment by yfarjoun** [comment]
@lh3 can you check my work?

@lbergelson you want to take a look here?

File: src/main/java/htsjdk/samtools/BAMRecord.java [hunk 1]
```diff
@@ -256,6 +283,55 @@ public Cigar getCigar() {
         return super.getCigar();
     }
 
+    private static boolean isSentinelCigar(final Cigar cigar, final int readLength) {
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/BAMRecord.java]: are SEQ allowed to be '*' when the cigar is populated? if so, this will not work in that case as it looks at the read-length for guidance. If folks think that this is too restrictive, I can just look for S and N operators and remove the readlength check.

Comment (lh3) [reply src/main/java/htsjdk/samtools/BAMRecord.java]: Perhaps change the condition to:
```java
return cigar.getCigarElement(0).getOperator() == CigarOperator.S &&
       (cigar.getCigarElement(0).getLength() == readLength || readLength == 0) &&
       cigar.getCigarElement(1).getOperator() == CigarOperator.N;
```

Comment (yfarjoun) [reply src/main/java/htsjdk/samtools/BAMRecord.java]: ah. I ignored this caveat. I thought that the sentinel is always xSyN

File: src/main/java/htsjdk/samtools/BAMRecordCodec.java [hunk 2]
```diff
@@ -174,6 +186,18 @@ public void encode(final SAMRecord alignment) {
...
+     * Create a "Sentinel" cigar that will be placed in BAM file when the actual cigar has more than 0xffff operator,
+     * which are not supported by the bam format. The actual cigar will be encoded and placed in the CG attribute.
+     * @param cigar actual cigar to create sentinel cigar for
+     * @return sentinel cigar xSyN with readLength (x) and referenceLength (y) matching the input cigar.
+     */
+    public static Cigar makeSentinelCigar(final Cigar cigar) {
+        return new Cigar(Arrays.asList(
+                new CigarElement(cigar.getReadLength(), CigarOperator.S),
+                new CigarElement(cigar.getReferenceLength(), CigarOperator.N)));
+    }
```
Comment (lh3) [inline src/main/java/htsjdk/samtools/BAMRecordCodec.java]: When writing to BAM, a CIGAR length is at most `(1<<28)-1`. Probably this has been taken care of somewhere else in the code. Just to make sure.

File: src/main/java/htsjdk/samtools/BAMRecord.java (hunk 1 above)
Comment (lh3) [reply src/main/java/htsjdk/samtools/BAMRecord.java]: Actually I am not familiar with this part. If SEQ='*', is readLength equal to 0? If so, your old initial code is ok.

File: src/main/java/htsjdk/samtools/BAMRecordCodec.java (hunk 2 above)
Comment (yfarjoun) [reply src/main/java/htsjdk/samtools/BAMRecordCodec.java]: where is this restriction coming from?

Comment (lh3) [reply src/main/java/htsjdk/samtools/BAMRecordCodec.java]: In BAM, the lower 4 bits for the operator and the higher 28 bits for the length, so if either query or the reference length in the alignment longer than `(1<<28)-1`, we will not be able to encode this in BAM. Encoding this in SAM is ok. I guess htsjdk probably has this check already somewhere.

**Comment by yfarjoun** [comment]
Got it. Did we say that in the spec pr?

File: src/main/java/htsjdk/samtools/BAMRecord.java (hunk 1 above)
Comment (lh3) [reply src/main/java/htsjdk/samtools/BAMRecord.java]: From @yfarjoun:

> Got it. Did we say that in the spec pr?

I don't know. This limitation is implicit in BAM (like chromosome lengths can't be longer than 2Gbp), but I agree it would be good to make it explicit anyway.

File: src/main/java/htsjdk/samtools/BAMRecordCodec.java (hunk 2 above)
Comment (lh3) [reply src/main/java/htsjdk/samtools/BAMRecordCodec.java]: From @yfarjoun:

Got it. Did we say that in the spec pr?

I don't know. This limitation is implicit in BAM (like chromosome lengths can't be longer than 2Gbp), but I agree it would be good to make it explicit anyway.

Comment (yfarjoun) [reply src/main/java/htsjdk/samtools/BAMRecordCodec.java]: This is an extra restriction for the case of a large cigar, not in general..since we will need to put the readlength into the cigar element and there's less room there. The read=length itself is encoded in a 32-bit int. I wonder if we can try to get around this restriction (e.g. if readlength>28-bit use (1<<28-)-1 in the cigar...)

Comment (lh3) [reply src/main/java/htsjdk/samtools/BAMRecordCodec.java]: In principle, yes. We may do that if Sanger people are happy with it.

Comment (yfarjoun) [reply src/main/java/htsjdk/samtools/BAMRecordCodec.java]: shouldn't the same restriction be placed on the reference length?

**Comment by yfarjoun** [comment]
@lh3 I made the fixes you asked for. a "*" read should have readlength 0, but even if not...it's not for this PR.

**Comment by yfarjoun** [comment]
@tfenne @jacarey @lbergelson: @lh3 approved this PR, could one of you check that it's OK with you as well?

File: src/test/java/htsjdk/samtools/BAMFileWriterTest.java [hunk 3]
```diff
@@ -60,50 +59,68 @@ private SAMRecordSetBuilder getRecordSetBuilder(final boolean sortForMe, final S
...
+        }
         if (presorted) { // If SAM text input was presorted, then we can compare SAM object to BAM object
             verifyBAMFile(samRecordSetBuilder, bamFile);
         }
+
+        File tempMetrics = File.createTempFile("CGTagTest", ".validation_metrics");
+        try (final SamReader samReader = SamReaderFactory.makeDefault().open(bamFile)) {
+
+            new SamFileValidator(new PrintWriter(new FileOutputStream(tempMetrics)),
+                    100).validateSamFileSummary(samReader, null);
```
Comment (lbergelson) [inline src/test/java/htsjdk/samtools/BAMFileWriterTest.java]: strange place for a linebreak here

File: src/test/java/htsjdk/samtools/BAMFileWriterTest.java [hunk 4]
```diff
@@ -60,50 +59,68 @@ private SAMRecordSetBuilder getRecordSetBuilder(final boolean sortForMe, final S
...
+            }
+        }
         if (presorted) { // If SAM text input was presorted, then we can compare SAM object to BAM object
             verifyBAMFile(samRecordSetBuilder, bamFile);
         }
+
+        File tempMetrics = File.createTempFile("CGTagTest", ".validation_metrics");
+        try (final SamReader samReader = SamReaderFactory.makeDefault().open(bamFile)) {
+
+            new SamFileValidator(new PrintWriter(new FileOutputStream(tempMetrics)),
```
Comment (lbergelson) [inline src/test/java/htsjdk/samtools/BAMFileWriterTest.java]: The `PrintWriter` could be created in the resource block for auto closing

File: src/test/java/htsjdk/samtools/BAMFileWriterTest.java [hunk 5]
```diff
@@ -229,4 +243,179 @@ public void testBAMReadBases() throws IOException {
...
+                {10_000},
+                {BAMRecord.MAX_CIGAR_OPERATORS - 1},
+                {BAMRecord.MAX_CIGAR_OPERATORS},
+                {BAMRecord.MAX_CIGAR_OPERATORS + 1},
+                {100_000},
+                {1_000_000}
+        };
+    }
+
+    final static CigarOperator operatorsToUse[] = new CigarOperator[]{CigarOperator.M, CigarOperator.D, CigarOperator.M, CigarOperator.I};
```
Comment (lbergelson) [inline src/test/java/htsjdk/samtools/BAMFileWriterTest.java]: private?  it's weird to put the `[]` on the name instead of the type

File: src/test/java/htsjdk/samtools/BAMFileWriterTest.java [hunk 6]
```diff
@@ -229,4 +243,179 @@ public void testBAMReadBases() throws IOException {
...
+
+    @Test(dataProvider = "longCigarsData")
+    public void testLongCigarsZerolengthRead(final int numOps) throws Exception {
+        final SAMRecordSetBuilder builder = new SAMRecordSetBuilder(true, SAMFileHeader.SortOrder.coordinate);
+        final Cigar cigar = Cigar.fromCigarOperators(getCigarOperatorsForTest(numOps));
+
+        final SAMRecord sam = builder.addFrag("frag1", 0, 1, false, false, cigar.toString(), null, 30);
+        sam.setReadBases(new byte[]{});
+        sam.setBaseQualityString("");
+        // in htsjdk only secondary alignments are allowed to have read-length zero (doens't validate otherwise)
```
Comment (lbergelson) [inline src/test/java/htsjdk/samtools/BAMFileWriterTest.java]: or supplementary?

File: src/test/java/htsjdk/samtools/BAMFileWriterTest.java [hunk 7]
```diff
@@ -229,4 +243,179 @@ public void testBAMReadBases() throws IOException {
...
+                bamWriter.addAlignment(record);
+        }
+
+        try (final SamReader reader = SamReaderFactory.makeDefault().validationStringency(ValidationStringency.SILENT).open(bamFile)) {
+            reader.iterator().forEachRemaining(SAMRecord::getCigar);
+        }
+    }
+
+    @Test(dataProvider = "longCigarsData")
+    public void testNoCGTagCigars(final int numOps) throws Exception {
```
Comment (lbergelson) [inline src/test/java/htsjdk/samtools/BAMFileWriterTest.java]: I think we should add tests to show that we don't have an incorrectly empty cigar if clearAttributes is called, and also test that we don't end up with an CG tag in the case where setCigar or setCigarString is called.

File: src/main/java/htsjdk/samtools/BAMRecord.java [hunk 8]
```diff
@@ -256,6 +284,56 @@ public Cigar getCigar() {
...
+        }
+
+        if (decodedCigar.getReadLength() != sentinelCigar.getReadLength()) {
+            throw new IllegalStateException(String.format(
+                    "Sentinel cigar and CG cigar should have the same read length. Found %d and %d.",
+                    sentinelCigar.getReadLength(),
+                    decodedCigar.getReadLength()));
+        }
+
+        setCigar(decodedCigar);
```
Comment (lbergelson) [inline src/main/java/htsjdk/samtools/BAMRecord.java]: Shouldn't this be a call to `initializeCigar` instead of `setCigar`? This will reset the indexing bin to null.

File: src/test/java/htsjdk/samtools/BAMFileWriterTest.java [hunk 9]
```diff
@@ -60,50 +59,68 @@ private SAMRecordSetBuilder getRecordSetBuilder(final boolean sortForMe, final S
...
+        try (final SamReader samReader = SamReaderFactory.makeDefault().open(bamFile)) {
+
+            new SamFileValidator(new PrintWriter(new FileOutputStream(tempMetrics)),
+                    100).validateSamFileSummary(samReader, null);
+        }
+
+        final MetricsFile<SamFileValidator.ValidationMetrics, String> validationMetrics = new MetricsFile<>();
+        validationMetrics.read(new FileReader(tempMetrics));
+
+        Assert.assertTrue(validationMetrics.getHistogram().get("ERROR:CG_TAG_FOUND_IN_ATTRIBUTES") == null);
```
Comment (lbergelson) [inline src/test/java/htsjdk/samtools/BAMFileWriterTest.java]: this can be `.assertNull`

File: src/test/java/htsjdk/samtools/BAMFileWriterTest.java [hunk 10]
```diff
@@ -60,50 +59,68 @@ private SAMRecordSetBuilder getRecordSetBuilder(final boolean sortForMe, final S
...
+                final SAMRecord samRecord1 = samIt1.next();
+                final SAMRecord samRecord2 = samIt2.next();
 
-            // Force reference index attributes to be populated
-            samRecord.getReferenceIndex();
-            bamRecord.getReferenceIndex();
-            samRecord.getMateReferenceIndex();
-            bamRecord.getMateReferenceIndex();
+                // SAMRecords don't have this set, so stuff it in there
+                samRecord2.setIndexingBin(samRecord1.getIndexingBin());
```
Comment (lbergelson) [inline src/test/java/htsjdk/samtools/BAMFileWriterTest.java]: I think this line is masking a bug that's resetting the indexing bin.  I'm not sure what to do about that in testing though.  Maybe we should have a test specifically for the indexing bin.

File: src/main/java/htsjdk/samtools/BAMRecord.java (hunk 8 above)
Comment (yfarjoun) [reply src/main/java/htsjdk/samtools/BAMRecord.java]: yes.....though now I'm annoyed that none of the tests caught this....what is bin used for ?

File: src/main/java/htsjdk/samtools/BAMRecord.java [hunk 11]
```diff
@@ -308,11 +386,17 @@ private void decodeAttributes() {
...
         }
+
         mAttributesDecoded = true;
         final int tagsOffset = readNameSize() + cigarSize() + basesSize() + qualsSize();
         final int tagsSize = mRestOfBinaryData.length - tagsOffset;
         final SAMBinaryTagAndValue attributes = BinaryTagCodec.readTags(mRestOfBinaryData, tagsOffset, tagsSize, getValidationStringency());
         setAttributes(attributes);
+
+        // if there's a CG tag, we should getCigar() so that the CG tag has a chance of turning into the CIGAR
+        if (hasAttribute(CG.name())) {
```
[the remark this answers is not in the document]
Comment (yfarjoun) [reply src/main/java/htsjdk/samtools/BAMRecord.java]: hmmm not sure I understand this....

File: src/main/java/htsjdk/samtools/BAMRecord.java (hunk 8 above)
Comment (lbergelson) [reply src/main/java/htsjdk/samtools/BAMRecord.java]: I think the tests explicitly don't test this.

Comment (lbergelson) [reply src/main/java/htsjdk/samtools/BAMRecord.java]: I'm honestly not certain how the bin is used.  I assume it's used in indexing?

File: src/main/java/htsjdk/samtools/BAMRecord.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (lbergelson) [reply src/main/java/htsjdk/samtools/BAMRecord.java]: The potential edge cases or the short from SAMTagUtils?

File: src/main/java/htsjdk/samtools/SAMRecord.java [hunk 12]
```diff
@@ -2048,7 +2048,7 @@ public int hashCode() {
             if (firstOnly) return ret;
         }
         // TODO(mccowan): Is this asking "is this the primary alignment"?
```
[the remark this answers is not in the document]
Comment (yfarjoun) [reply src/main/java/htsjdk/samtools/SAMRecord.java]: not sure...reverting this, as I'm not sure it's right anymore...

File: src/test/java/htsjdk/samtools/BAMFileWriterTest.java (hunk 6 above)
Comment (yfarjoun) [reply src/test/java/htsjdk/samtools/BAMFileWriterTest.java]: nope.

File: src/test/java/htsjdk/samtools/BAMFileWriterTest.java (hunk 7 above)
Comment (yfarjoun) [reply src/test/java/htsjdk/samtools/BAMFileWriterTest.java]: lets discuss these cases in person.

**Comment by yfarjoun** [comment]
@lbergelson thanks for the review. the tests for the non-null bin catch the bug you identified regarding setCigar vs. initializeCigar...thanks!

**Comment by yfarjoun** [comment]
had to rebase to get rid of the FTP errors...sorry.

**Comment by yfarjoun** [comment]
oops. didn't mean to close this.

**Comment by yfarjoun** [comment]
I couldn't find the typo....

File: src/main/java/htsjdk/samtools/BAMRecord.java [hunk 13]
```diff
@@ -284,15 +284,27 @@ public Cigar getCigar() {
...
+                (cigar.getCigarElement(0).getLength() == readLength || readLength == 0) ;
     }
 
-    private void extractCigarFromCGAttribute(final Cigar sentinelCigar) {
-        final int[] cigarFromCG = (int[]) getAttribute(CG.name());
+
+    /**
+     * Long cigars (with more than 64K operators) cannot be encoded into BAM. Instead a sentinel cigar is
+     * placed as a placeholder, and the actual cigar is placed in the CG tag. This method
+     * extracts the CIGAR from the CG tag and places is into the (in memory) cigar.
```
Comment (lbergelson) [inline src/main/java/htsjdk/samtools/BAMRecord.java]: here:  is -> it

I thought I'd commented here but somehow I didn't post it.

Comment (lbergelson) [reply src/main/java/htsjdk/samtools/BAMRecord.java]: Sorry, I didn't mean it to be a game of wordhunt

**Comment by yfarjoun** [comment]
:party: long cigars are supported in htsjdk!!!
